text/0000-project-error-handling.md

```diff
@@ -0,0 +1,121 @@
+
+### Come to a consensus on current best practices
+
+Here is a tenative starting point, subject to change:
+
+- Use `Result` and `Error` types for recoverable errors.
+- Use `panic` for unrecoverable errors.
+- Impl `Error` for error types that may need to be reported to a human or be composed with other errors.
+- Use enums for types representing multiple failure cases that may need to be handled.
+    - For libraries, oftentimes you want to support both reporting and handling so you impl `Error` on a possibly non-exhaustive enum.
+- Error kind pattern for associating context with every enum variant without including the member in every enum variant.
+- Convert to a reporting type when the error is no longer expected to be handled beyond reporting e.g. `anyhow::Error` or `eyre::Report` or when trait object + downcast error handling is preferable.
+- Recommend `Box`ing concrete error types when stack size is an issue rather than `Box`ing and converting to `dyn Error`s.
+- What is the consensus on handling `dyn Error`s? Should it be encouraged or discouraged? Should we look into making `Box<dyn Error...>` impl Error?
+
+
+### Communicate current best practices
+
+- Document the concensus.
+- Communicate plan for future changes to error handling, and the libraries that future changes are being based off of.
+- Produce learning resources related to current best practices.
+    - New chapters in the book?
+
+### Evaluate options for error reporting type aka better `Box<dyn Error>`
+
+- Survey the current libraries in the ecosystem:
+    - `anyhow`
+    - `eyre`
+- Evaluate value of features including:
+    - Single word width on stack
+    - Error wrapping with display types and with special downcast support.
+    - report hook + configurable `dyn ReportHandler` type for custom report formats and content, similar to panic handler but for errors.
+    - libcore compatibility.
+
+### Consolidate ecosystem by merging best practice crates into std
+
+- Provide a derive macro for `Error` in std.
+- Stabilize the `Backtrace` type but possibly not `fn backtrace` on the Error trait.
+    - Provide necessary API on `Backtrace` to support crates like `color-backtrace`.
+- Move error to core .
+    - Depends on generic member access.
+    - Requires resolving downcast dependency on `Box` and blocking the stabilization of `fn backtrace`.
+- Potentially stabilize an error Reporting type based on `anyhow` and `eyre` now that they're close to having identical feature sets.
+
+### Add missing features
+
+- Generic member access on the `Error` trait.
+- `Error` return traces:
+    - Depends on specialization and generic member access.
+- Fix rough corners around reporting errors and `Termination`.
+
+## Non Goals
+
+- This group should not be involved in design discussions for the `Try` trait, `try` blocks, or `try` fns.
```

Comment:
"Involved" doesn't appear to be the right word. The way this is phrased, it sounds as if **members** of this group are not allowed to participate in the discussion.

Reply:
```suggestion
- This group should not be managing in design discussions for the `Try` trait, `try` blocks, or `try` fns.
```

Hows this?

Reply:
Another option: add a parenthetical to the effect of "(Not that people can't participate in both this group and in `try` stuff)"